API: more permissive conversion to StringDtype

topper-123 opened the pull request:
- [x] closes #31204
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is a proposal to make using ``StringDtype`` more permissive and be usable inplace of ``dtype=str``.

ATM converting to StringDtype will only accept arrays that are ``str`` already, meaning you will often have to use ``astype(str).astype("string")`` to be sure not to get errors, which can be tedious. For example these fail in master and work in this PR:

```python
>>> pd.Series([1,2, np.nan], dtype="string")
0       1
1       2
2    <NA>
dtype: string
>>> pd.array([1,2, np.nan], dtype="string")
<StringArray>
['1', '2', <NA>]
Length: 3, dtype: string
>>> pd.Series([1,2, np.nan]).astype("string")
0     1.0
1     2.0
2    <NA>
dtype: string
>>> pd.Series([1,2, np.nan], dtype="Int64").astype("string")
0       1
1       2
2    <NA>
dtype: string
```

etc. now work. Previously the above all gave errors.

## The proposed solution:
``ExtensionArray._from_sequence`` is in master explicitly stated to expect a sequence of scalars of the *correct* type. This makes it not usable for type conversion.

I've therefore added a new ``ExtensionArray._from_sequence_of_any_type`` that accepts scalars of any type and may change the scalars to the correct type before passing them on to ``_from_sequence``. Currently it just routes everything through to ``ExtensionArray._from_sequence``, except in ``StringArray._from_sequence_of_any_type``, where it massages the input scalars into strings before passing them on.

Obviously tests and doc updates are still missing, but I would appreciate feedback if this solution looks ok and I'll add tests and doc updates if this is ok.

TomAugspurger commented:
https://github.com/pandas-dev/pandas/issues/33254 has a discussion about the appropriate strictness of `_from_sequence`.

topper-123 commented:
Ther doctest failures are very strange, it doesn't make sense that they're failing in my PR, but not in other PRs...

From the discussion in #33254 it looks like a dedicated ``_from_sequence_of_any_type`` would not be the consensus approach, but maybe push its content to ``_from_sequence`` and maybe add a ``_from_scalars``, but it's doesnt seem like the issue is resolved.

any comments on the current PR?

jorisvandenbossche commented:
The implementation itself looks good to me. 
Awaiting resolution of the discussion in #33254, I think in principle we can already add this behaviour to ``_from_sequence``, it would at least fix the conversion issues we have right now.

topper-123 commented:
I've updated the PR:
* Moved the new functionality to ``StringArray._from_sequence``
* added tests to ``BaseCastingTests.test_astype_string``
* made changes to make the tests pass
* updated docs.
* Added a ``dtype`` argument to ``Series.combine``.

The last bullet point is because of a situation where you combine two ``string`` Series and want the result to be in another extension dtype is very difficult to in a inferred way. I think it's better and safer in those cases to just supply the desired dtype.

topper-123 commented on pandas/core/arrays/sparse/dtype.py:
```diff
@@ -320,7 +320,7 @@ def update_dtype(self, dtype):
 
         if not isinstance(dtype, cls):
-            fill_value = astype_nansafe(np.array(self.fill_value), dtype).item()
+            fill_value = astype_nansafe(np.array([self.fill_value]), dtype)[0]
```
ExtensionArrays don't have an ``.item`` method.

jorisvandenbossche replied:
Do you remember why this was needed? (Sparse currently doesn't support using EAs under the hood, so in principle, we should never get an extension dtype here)

topper-123 replied:
The situation that fails if we do .item is:

dtype is a StringDtype and cls is a SparseDtype. I'm not very familiar with sparse arrays, but it seems reasonable that string arrays can be converted to sparse arrays?

jorisvandenbossche replied:
> it seems reasonable that string arrays can be converted to sparse arrays?

In principle, yes, but in practice they don't support EAs as underlying values (the `SparseArray.__init__` converts any input to a numpy array). So I think this should raise an informative error when `dtype` is an extensiondtype and not a numpy dtype.

topper-123 committed:
API: More permissive conversion to StringDtype

topper-123 committed:
clean up rst and doc strings

topper-123 commented on pandas/tests/extension/base/methods.py:
```diff
@@ -172,15 +172,16 @@ def test_combine_le(self, data_repeated):
         s1 = pd.Series(orig_data1)
         s2 = pd.Series(orig_data2)
-        result = s1.combine(s2, lambda x1, x2: x1 <= x2)
+        result = s1.combine(s2, lambda x1, x2: x1 <= x2, dtype="boolean")
```
Inferring to a different dtype than the original is quite difficult, and better to be explicit in those cases IMO.

jorisvandenbossche replied:
Although I agree in general (certainly when using `combine` for constructing the expected result), but is this change required in this specific case? (I would assume nothing changed related to this?) Because I might still be good to test the case *without* specifying the dtype explicitly, as that is public API.

topper-123 replied:
The problem here is that almost anything can be converted to a string. I've made a change though. It's inferring dtype, so not ideal performance-wise in the case where the series is a StringDtype and the result might not be.

If the user adds dtype manually, the performance is ok.

jorisvandenbossche commented on pandas/core/arrays/string_.py:
```diff
@@ -104,6 +104,11 @@ class StringArray(PandasArray):
+    .. versionchanged:: 1.1.0
+
+        ``StringArray`` allow non-string input values, but will always convert the
+        values  to strings. (Before Pandas 1.1 non-string values were not allowed).
```
This is not really a change in the `StringArray` constructor (that still disallows non-strings), so I would leave out the versionchanged note here

topper-123 replied:
Yeah, deleted.

jorisvandenbossche commented on pandas/core/arrays/string_.py:
```diff
@@ -152,13 +157,13 @@ class StringArray(PandasArray):
     Length: 4, dtype: string
 
-    Unlike ``object`` dtype arrays, ``StringArray`` doesn't allow non-string
-    values.
```
I think it is important to keep the original intent of this part, in some way. As it was meant to explain that the StringArray will only *contain* strings (as opposed to object dtype, which is not strict here). This could maybe also shown with setting a non-string value (`arr[0] = 1`) or so.

jorisvandenbossche replied:
How you changed it still doesn't explain the original intent, as mentioned above, IMO. I would explain both (only strings allowed + showing the automatic conversion)

topper-123 replied:
Ok, I've tried it differently.

jorisvandenbossche commented on pandas/core/series.py:
```diff
@@ -2682,6 +2682,11 @@ def combine(self, other, func, fill_value=None) -> "Series":
             appropriate NaN value for the underlying dtype of the Series.
+        dtype : str, numpy.dtype, or ExtensionDtype, optional
+            Data type for the output Series. If not specified, this will be
+            inferred from the combined data.
```
+1 on adding such a keyword!

jorisvandenbossche commented on pandas/tests/extension/base/casting.py:
```diff
@@ -36,6 +36,11 @@ def test_astype_str(self, data):
 
+    def test_astype_string(self, data):
+        result = pd.Series(data[:5]).astype("string")
+        expected = pd.Series(data[:5].astype("string"))
```
Can you create the expected without relying on the EA method (eg doing the `astype(str)` workaround) ?

topper-123 replied:
Ok.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -89,6 +115,7 @@ Other enhancements
 - :meth:`DataFrame.sample` will now also allow array-like and BitGenerator objects to be passed to ``random_state`` as seeds (:issue:`32503`)
 - :meth:`MultiIndex.union` will now raise `RuntimeWarning` if the object inside are unsortable, pass `sort=False` to suppress this warning (:issue:`33015`)
 - :class:`Series.dt` and :class:`DatatimeIndex` now have an `isocalendar` accessor that returns a :class:`DataFrame` with year, week, and day calculated according to the ISO 8601 calendar (:issue:`33206`).
+- :meth:`Series.combine` has gained a ``dtype`` argument. If supplied, the combined series will get that dtype (:issue:`33465`)
```
not averse to this, but can we do as a separate change?

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -431,6 +431,11 @@ def astype(self, dtype, copy=True):
         array : ndarray
             NumPy ndarray with 'dtype' for its dtype.
         """
+        from pandas.core.arrays.string_ import StringDtype
```
so we atcually *ever* hit this base type? I think we override this everywhere

topper-123 replied:
floats arrays are PandasArrays, and those get their astype from ExtensionsArray.

```python
>>> pd.array([1.5, 2.5])
<PandasArray>
[1.5, 2.5]
Length: 2, dtype: float64
```

Allowing any ExtensionArray by default to convert to StringArray seems reasonable to me (and if subclassers don't want that, they can make their own astype implementation disallowing StringArrays).

jreback replied:
kk, can you add a commment to this effect here (agree with your statement), but comment for future readers

jorisvandenbossche replied:
> can you add a commment to this effect here

All methods in this base class are there for subclasses to (potentially) use, so I don't think a comment about that is needed 
(a comment about always being able to astype to string dtype is fine though)

jreback replied:
i understand I would move the impl that is currently in Decimal to here as it correctly handles the astype from the *same* type (whereas this one will coerce to a numpy array)

jreback commented on pandas/tests/extension/base/casting.py:
```diff
@@ -36,6 +36,11 @@ def test_astype_str(self, data):
         self.assert_series_equal(result, expected)
 
+    def test_astype_string(self, data):
+        result = pd.Series(data[:5]).astype("string")
```
can you give the issue ref number here

topper-123 replied:
Ok.

jreback commented on pandas/tests/extension/base/methods.py:
```diff
@@ -172,15 +172,16 @@ def test_combine_le(self, data_repeated):
+        result = s1.combine(s2, lambda x1, x2: x1 <= x2, dtype="boolean")
         expected = pd.Series(
-            [a <= b for (a, b) in zip(list(orig_data1), list(orig_data2))]
+            [a <= b for (a, b) in zip(list(orig_data1), list(orig_data2))],
```
what happens if we *don't* specify dtype? can you add examples for this as well (its possible we should also raise in some cases)

topper-123 replied:
I've made a change to infer, i.e. give the same result as before.

jreback commented on pandas/tests/extension/decimal/array.py:
```diff
@@ -130,8 +131,13 @@ def copy(self):
     def astype(self, dtype, copy=True):
+        from pandas.core.arrays.string_ import StringDtype
+
+        dtype = pandas_dtype(dtype)
```
I think this here should be the array/base.py impl.

jreback replied:
can you do this

jorisvandenbossche replied:
> I think this here should be the array/base.py impl.

The base class version already does this?

topper-123 replied:
I've replaced that with a ``super``-call.

jreback replied:
see my comments above

jorisvandenbossche replied:
The decimal astype implementation needs to handle the "context", so can't be fully moved to the base class (@topper-123 already moved what can be moved and calls `super()` in the end.

jorisvandenbossche reviewed (commented):
Thanks for this, @topper-123 (and sorry for the slow review). As I said in https://github.com/pandas-dev/pandas/issues/33254, I think we can fix this bug regardless of the progress of the bigger discussion in that issue.

topper-123 committed:
update

jorisvandenbossche commented:
@topper-123 workflow related question: can you please not force-push, but just add new commits (and merge master for updating with master)? That makes it a lot easier to review what has been updated since a previous review (and ensures github's UI to help with this actually works)

topper-123 committed:
changes

topper-123 commented:
Ok, no problem,  I’ll stop force-pushing. I thought that made the commits more tidy and easier to read, but if that’s wrong, I’ll make regular commits in the future.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -13,6 +13,32 @@ including other versions of pandas.
 
+.. _whatsnew_110.astype_string:
+
+All dtypes can now be converted to ``StringDtype``
```
extension dtypes

jorisvandenbossche replied:
No, all dtypes (this not only about extension dtypes)

topper-123 replied:
I'll leave it as it it,

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -13,6 +13,32 @@ including other versions of pandas.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously, declaring or converting to :class:`StringDtype` was in general only possible if the data was already only ``str`` or nan-like.
+For example:
```
can you use the 

Previous

Current

way of formatting (see other notes)

jorisvandenbossche replied:
I would personally just leave out the "previous". Saying that it didn't work before, and only showing the new feature seems sufficient to me.

jreback replied:
please use the prescribed format, its a standard that we have and consistency is key

jorisvandenbossche replied:
It's only a standard when there is actually a "Previous" to show. For bugs, we also don't show the error message you got before it was fixed. I personally see the fact that you couldn't do `astype("string")` to convert to strings as a bug (or not yet implemented feature, given how recent StringDtype is)

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -13,6 +13,32 @@ including other versions of pandas.
+All dtypes can now be converted to ``StringDtype``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously, declaring or converting to :class:`StringDtype` was in general only possible if the data was already only ``str`` or nan-like.
```
list the closed issues here at the end of the first statement

topper-123 replied:
Ok.

jreback commented on pandas/core/arrays/string_.py:
```diff
@@ -203,10 +207,14 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+        if has_nans and result is scalars:
+            # force a copy now, if we haven't already
+            result = result.copy()
+        # convert to str, then to object to avoid dtype like '<U3', then insert na_value
```
put a blank line

topper-123 replied:
Ok.

jreback commented on pandas/core/series.py:
```diff
@@ -2768,6 +2768,10 @@ def combine(self, other, func, fill_value=None) -> "Series":
         if is_categorical_dtype(self.dtype):
             pass
         elif is_extension_array_dtype(self.dtype):
+            # Everything can be be converted to strings, but we may not want to convert
```
realy don't add things here, if you need to modify maybe_cast_to_extesion_array then its ok

this becomes hugely complicated to follow the path otherwise

jorisvandenbossche replied:
Yes, I would move this into `maybe_cast_to_extension_array` (we will be able to remove this once the "strict" `from_scalars` is implemented, and having the logic in `maybe_cast_to_extension_array` will make it easier to not forget to update that)

topper-123 replied:
Ok, found a way to avoid changing series.py at all.

topper-123 committed:
Change according to comments

topper-123 committed:
Small clean-up

topper-123 committed:
failed doctest

topper-123 committed:
update doc string

topper-123 commented:
Updated.

jreback commented on doc/source/user_guide/text.rst:
```diff
@@ -63,6 +63,27 @@ Or ``astype`` after the ``Series`` or ``DataFrame`` is created
+
+or convert from existing pandas data:
+
+   s1 = pd.Series([1,2, np.nan], dtype="Int64")
```
this won't render (needs an ipython block here as well)

topper-123 replied:
Fixed.

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -442,17 +442,20 @@ def astype(self, dtype, copy: bool = True) -> ArrayLike:
-            return BooleanArray(result, mask=self._mask, copy=False)
+            return dtype.construct_array_type()(result, mask=self._mask, copy=False)
+        elif isinstance(dtype, StringDtype):
+            return dtype.construct_array_type()._from_sequence(self, copy=False)
```
why is mask not passed in? and why is this using _from_sequence? e.g. this construction is *different* than from Integer/boolean

jreback replied:
we should either *always* use _from_sequence of never, mixing these is *very* confusing.

jorisvandenbossche replied:
That is because the BooleanArray constructor takes a mask, and the StringArray not. StringArray is not (yet) a masked array, so has a slightly different API.

(we need to revisit the full astype implementation anyway: https://github.com/pandas-dev/pandas/issues/22384)

jreback replied:
ok it would be great to do this next, as we are already creating a fair amount of tech debt here

jorisvandenbossche replied:
> to do this next

Making StringArray into a masked array, or the astype refactor? 
(the astype refactor actually first needs a concrete proposal, but I am planning to look into it after concat)

jreback replied:
> Making StringArray into a masked array

I thought this was *already* the case, I guess not.

jorisvandenbossche replied:
It used pd.NA as missing value, but it is not implemented as a masked array

jorisvandenbossche commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -13,6 +13,24 @@ including other versions of pandas.
+Previously, declaring or converting to :class:`StringDtype` was in general only possible if the data was already only ``str`` or nan-like (:issue:`31204`).
+:class:`StringDtype` now works in all situations where ``astype(str)`` or ``dtype=str`` work:
+
+For example the below now work:
```
```suggestion
For example, the below now works:
```

topper-123 committed:
minor updates

topper-123 committed:
lint

jreback commented:
thanks @topper-123 very nice!
